BlueMapCore/src/main/java/de/bluecolored/bluemap/core/storage/s3/S3Storage.java

```diff
@@ -0,0 +1,342 @@
+package de.bluecolored.bluemap.core.storage.s3;
+
+import com.flowpowered.math.vector.Vector2i;
+import de.bluecolored.bluemap.core.storage.*;
+import de.bluecolored.bluemap.core.util.OnCloseOutputStream;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.*;
+
+import java.io.*;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class S3Storage extends Storage {
+    private boolean closed = false;
+    private final S3Client client;
+    private final Compression hiresCompression;
+    private final String bucket;
+    public S3Storage(S3StorageSettings settings) {
+        AwsSessionCredentials awsCreds = AwsSessionCredentials.create(settings.getAccessKey(), settings.getSecretKey(), "");
+        var builder = S3Client
+                .builder()
+                .credentialsProvider(StaticCredentialsProvider.create(awsCreds))
+                .region(Region.of(settings.getRegion()));
+        if (settings.getEndpoint().isPresent()) {
+            builder = builder.endpointOverride(URI.create(settings.getEndpoint().get()));
+        }
+        this.client = builder.build();
+        this.hiresCompression = settings.getCompression();
+        this.bucket = settings.getBucket();
+    }
+    @Override
+    public void initialize() throws IOException {}
+
+    @Override
+    public OutputStream writeMapTile(String mapId, int lod, Vector2i tile) throws IOException {
+        Compression compression = lod == 0 ? this.hiresCompression : Compression.NONE;
+        String path = getFilePath(mapId, lod, tile);
+
+        OutputStream pipeOut = makeUploadStream(path, getFileMime(lod));
+        return new BufferedOutputStream(compression.compress(pipeOut));
+    }
+
+    @Override
+    public Optional<CompressedInputStream> readMapTile(String mapId, int lod, Vector2i tile) throws IOException {
+        Compression compression = lod == 0 ? this.hiresCompression : Compression.NONE;
+        String path = getFilePath(mapId, lod, tile);
+        try {
+            InputStream is = client.getObject(
+                    GetObjectRequest
+                            .builder()
+                            .bucket(bucket)
+                            .key(path)
+                            .build()
+            );
+            return Optional.of(new CompressedInputStream(is, compression));
+        } catch (Exception e) {
+            if (e instanceof NoSuchKeyException) {
+                return Optional.empty();
+            }
+            throw new IOException(e);
+        }
```

Comment:
```suggestion
        } catch (NoSuchKeyException e) {
            return Optional.empty();
        } catch (SdkException e) {
            throw new IOException(e);
        }
```
Same for all the other caches :)

Reply:
remnant from me using minio at first and switching to aws-sdk only after it failed to build, yeah